kythe/go/serving/pipeline/beam.go

```diff
@@ -89,6 +93,54 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
 	return FromNodes(s, nodes.FromEntries(s, entries))
 }
 
+// CrossReferences returns a Kythe file decorations table derived from the Kythe
+// input graph.  The beam.PCollections have elements of type
+// KV<string, *srvpb.PagedCrossReferences> and
+// KV<string, *srvpb.PagedCrossReferences_Page>, respectively.
+func (k *KytheBeam) CrossReferences() (sets, pages beam.PCollection) {
+	s := k.s.Scope("CrossReferences")
+	refs := beam.GroupByKey(s, beam.ParDo(s, keyRef, k.References()))
+	// TODO(schroederc): related nodes
+	// TODO(schroederc): callers
+	// TODO(schroederc): MarkedSource
+	// TODO(schroederc): source_node
+	return beam.ParDo2(s, groupCrossRefs, refs)
+}
+
+func groupCrossRefs(key *spb.VName, refStream func(**ppb.Reference) bool, emitSet func(string, *srvpb.PagedCrossReferences), emitPage func(string, *srvpb.PagedCrossReferences_Page)) {
+	set := &srvpb.PagedCrossReferences{SourceTicket: kytheuri.ToString(key)}
+	// TODO(schroederc): add paging
+
+	groups := make(map[string]*srvpb.PagedCrossReferences_Group)
+
+	var ref *ppb.Reference
+	for refStream(&ref) {
+		kind := refKind(ref)
+		g, ok := groups[kind]
+		if !ok {
+			g = &srvpb.PagedCrossReferences_Group{Kind: kind}
+			groups[kind] = g
+			set.Group = append(set.Group, g)
+		}
+		g.Anchor = append(g.Anchor, ref.Anchor)
+		set.TotalReferences++
```

Comment:
what is total_references actually used for?  that is, why is it a field in the proto, and not just an internal value used by the pipeline to decide when to give up on a particular record?

Reply:
I'm deprecating it; the field isn't useful anymore.